File: src/advent_of_code/2022/day08.clj
(ns advent-of-code.2022.day08
  "https://adventofcode.com/2022/day/8"
  (:require
   [clojure.string :as str]))

(def input-filename "inputs/2022/day08.txt")

(def tree-grid (->> (slurp input-filename)
                    (str/split-lines)
                    (mapv (fn [line-s] (->> (str/split line-s #"")
                                            (mapv #(Integer/parseInt %)))))))

(defn transpose 
  "essentially swaps rows and columns"
  [grid] (apply map list grid))

(defn reverse-rows 
  "reverses direction of all the rows"
  [grid] (map reverse grid))

;; following functions take a `transform-fn` which is applied in the direction corresponding to the name of the function
;; from-left-matrix for example applies transform-fn to all rows
;;   from-rght-matrix does the same but reverses the row before passing it to transform-fn
;; from-top-matrix for example applies transform-fn to all columns
;;  from-botton-matrix does the same but reverses the column before passing it to transform-fn

;; the reason it was done in such a way is so that one function (say a function for calculating the visibility of tree or 
;; a function for calculating number of tress visible) could be coded such that it calculates required value roww-se from the left 
;; and it can then be used for all 4 directions 

(defn from-left-matrix [transform-fn]
  (->> tree-grid
       (map transform-fn)))

(defn from-top-matrix [transform-fn]
  (->> tree-grid
       transpose
       (map transform-fn)
       transpose))

(defn from-right-matrix [transform-fn]
  (->> tree-grid
       reverse-rows
       (map transform-fn)
       reverse-rows))

(defn from-bottom-matrix [transform-fn]
  (->> tree-grid
       transpose
       reverse-rows
       (map transform-fn)
       reverse-rows
       transpose))

;;;; PART 1

;; function which traverses a row and returns true for the trees that are visible from the left
(defn tree-row-visible-from-left-bools [tree-row]
  (:acc (reduce (fn [{:keys [acc max]} tree-h]
                  (let [new-max? (< max tree-h)]
                    {:acc (conj acc new-max?)
                     :max (if new-max? tree-h max)}))
                {:max -1 ; since there are trees of height 0
                 :acc []}
                tree-row)))

(assert (= (tree-row-visible-from-left-bools '(2 5 5 1 2))
           [true true false false false]))

;; bool lists for each row
(def bool-matrix-visible-trees
  (->> tree-row-visible-from-left-bools
       ((juxt from-left-matrix from-top-matrix from-right-matrix from-bottom-matrix))
       (apply map vector)
       (map (fn [bool-lists-for-a-row] (reduce (fn [acc row]
                                                 (map #(or %1 %2) acc row))
                                               bool-lists-for-a-row)))))

(->> bool-matrix-visible-trees
     flatten
     (filter true?)
     count)
;; part 1: 1662

;; PART 2
;; fuck it. I'm switching over to using indexes 

;; function which traverses a row and returns numbers of trees visible to the left
(defn num-trees-visible-to-left [tree-row i]
  (let [val-i (nth tree-row i)]
    (loop [m (dec i)]
      (if (< m 0)
        i
        (if (<= val-i (nth tree-row m))
          (- i m)
          (recur (dec m)))))))

(defn num-trees-visible-to-left-fn-for-whole-row [tree-row] (map (partial num-trees-visible-to-left tree-row) (range (count tree-row))))

(def matrix-scenic-scores
  (->> num-trees-visible-to-left-fn-for-whole-row
       ((juxt from-left-matrix from-top-matrix from-right-matrix from-bottom-matrix))
       (apply map vector)
       (map (fn [num-tress-visible-lists-for-a-row]
              (reduce (fn [acc row]
                        (map #(* %1 %2) acc row))
                      ;; contains 4 elems, one for each direction
                      num-tress-visible-lists-for-a-row)))))

(->> matrix-scenic-scores
     flatten
     (apply max))
;; part 2: 537600

File: python/2022/day08.py
"""Day 8: Treetop Tree House - https://adventofcode.com/2022/day/8"""

INPUT_FILE = "inputs/2022/day08.txt"


def parse_input(filename: str) -> list[list[int]]:
    """Parse input into 2D grid of tree heights."""
    with open(filename) as f:
        return [[int(c) for c in line.strip()] for line in f]


def transpose(grid: list[list[int]]) -> list[list[int]]:
    """Transpose a 2D grid (swap rows and columns)."""
    return [list(row) for row in zip(*grid)]


def reverse_rows(grid: list[list[int]]) -> list[list[int]]:
    """Reverse each row in the grid."""
    return [row[::-1] for row in grid]


def visible_from_left(row: list[int]) -> list[bool]:
    """For each tree, check if it's visible from the left."""
    result = []
    max_height = -1
    for height in row:
        if height > max_height:
            result.append(True)
            max_height = height
        else:
            result.append(False)
    return result


def viewing_distance_left(row: list[int], idx: int) -> int:
    """Count trees visible to the left from position idx."""
    height = row[idx]
    count = 0
    for i in range(idx - 1, -1, -1):
        count += 1
        if row[i] >= height:
            break
    return count


def compute_visibility_matrix(grid: list[list[int]]) -> list[list[bool]]:
    """Compute visibility from all 4 directions."""
    rows = len(grid)
    cols = len(grid[0])

    # Initialize result matrix
    visible = [[False] * cols for _ in range(rows)]

    # From left
    for r, row in enumerate(grid):
        vis = visible_from_left(row)
        for c in range(cols):
            visible[r][c] = visible[r][c] or vis[c]

    # From right
    reversed_grid = reverse_rows(grid)
    for r, row in enumerate(reversed_grid):
        vis = visible_from_left(row)
        for c in range(cols):
            visible[r][cols - 1 - c] = visible[r][cols - 1 - c] or vis[c]

    # From top
    transposed = transpose(grid)
    for c, col in enumerate(transposed):
        vis = visible_from_left(col)
        for r in range(rows):
            visible[r][c] = visible[r][c] or vis[r]

    # From bottom
    transposed_reversed = reverse_rows(transposed)
    for c, col in enumerate(transposed_reversed):
        vis = visible_from_left(col)
        for r in range(rows):
            visible[rows - 1 - r][c] = visible[rows - 1 - r][c] or vis[r]

    return visible


def compute_scenic_scores(grid: list[list[int]]) -> list[list[int]]:
    """Compute scenic score for each tree."""
    rows = len(grid)
    cols = len(grid[0])
    scores = [[1] * cols for _ in range(rows)]

    # Left viewing distance
    for r, row in enumerate(grid):
        for c in range(cols):
            scores[r][c] *= viewing_distance_left(row, c)

    # Right viewing distance - iterate on reversed row, map back to original
    reversed_grid = reverse_rows(grid)
    for r, row in enumerate(reversed_grid):
        for c in range(cols):
            # c is position in reversed row, (cols-1-c) is position in original
            scores[r][cols - 1 - c] *= viewing_distance_left(row, c)

    # Top viewing distance
    transposed = transpose(grid)
    for c, col in enumerate(transposed):
        for r in range(rows):
            scores[r][c] *= viewing_distance_left(col, r)

    # Bottom viewing distance - iterate on reversed column, map back to original
    transposed_reversed = reverse_rows(transposed)
    for c, col in enumerate(transposed_reversed):
        for r in range(rows):
            # r is position in reversed column, (rows-1-r) is position in original
            scores[rows - 1 - r][c] *= viewing_distance_left(col, r)

    return scores


def part1(grid: list[list[int]]) -> int:
    """Count trees visible from outside."""
    visibility = compute_visibility_matrix(grid)
    return sum(1 for row in visibility for visible in row if visible)


def part2(grid: list[list[int]]) -> int:
    """Find highest scenic score."""
    scores = compute_scenic_scores(grid)
    return max(max(row) for row in scores)


def main():
    grid = parse_input(INPUT_FILE)

    answer1 = part1(grid)
    answer2 = part2(grid)

    print(f"Part 1: {answer1}")  # 1662
    print(f"Part 2: {answer2}")  # 537600

    return answer1, answer2


if __name__ == "__main__":
    main()
